Libraries/LibJS/Runtime/Object.cpp

```diff
@@ -228,33 +279,37 @@ bool Object::put_own_property(Object& this_object, const FlyString& property_nam
         ASSERT(metadata.has_value());
     }
 
-    if (!new_property && mode == PutOwnPropertyMode::DefineProperty && !(metadata.value().attributes & Attribute::Configurable) && attributes != metadata.value().attributes) {
-        dbg() << "Disallow reconfig of non-configurable property";
-        if (throw_exceptions)
-            interpreter().throw_exception<TypeError>(String::format("Cannot redefine property '%s'", property_name.characters()));
-        return false;
+    auto value_here = m_storage[metadata.value().offset];
+    if (!new_property && mode == PutOwnPropertyMode::DefineProperty && !(metadata.value().attributes & Attribute::Configurable)) {
+        if (attributes != metadata.value().attributes || value_here.is_getter_setter_pair() != value.is_getter_setter_pair()) {
+            dbg() << "Disallow reconfig of non-configurable property";
+            if (throw_exceptions)
+                interpreter().throw_exception<TypeError>(String::format("Cannot change attributes of non-configurable property '%s'", property_name.characters()));
+            return false;
+        }
     }
 
+    // FIXME: Transitions from accessor to data, or vice versa
     if (mode == PutOwnPropertyMode::DefineProperty && attributes != metadata.value().attributes) {
         if (m_shape->is_unique()) {
             m_shape->reconfigure_property_in_unique_shape(property_name, attributes);
         } else {
             set_shape(*m_shape->create_configure_transition(property_name, attributes));
         }
```

Comment:
I wasn't quite sure how to enforce a shape transition here, where an object property is changing from a data descriptor to an accessor descriptor or vice versa (because the attributes don't necessary have to change). In the first iteration of the PR, I had Shapes and TransitionKeys take a `bool is_data_property`, which I could then change in the `create_configure_transition` call to make a transition. Or to save space, I could put it on the `u8 attributes`. I can re-implement that behavior if that's the best way.

Reply:
Hmm. I think the most straightforward way to do this would be adding two (engine-internal) attributes: something like `HasGet` and `HasSet`. Then we can just piggyback on the existing "configure" transitions. Does that make sense to you?